### PR TITLE
Implement Dimming actuator

### PR DESCRIPTION
**Pull request by derjoerg**
This implements the Dimming actuator

### CODE REVIEW

**Comment by derjoerg** [comment]
I don't get it how to test the set_brightness() function

**Comment by kingsleyadam** [comment]
> This implements the Dimming actuator

Have you been able to run tests locally? That'll help a lot in triaging the tests.

Setup a Python virtual environment and install the PyPi package `pip install .` while in the root directory.

Install Python dev dependencies `pip install src/abbfreeathome/dev-requirements.txt`

Then run the following to run tests.

```
pytest \                                                                                                                                                                                                                                                
  --doctest-modules \
  --junitxml=junit/test-results.xml \
  --cov=abbfreeathome \
  --cov-report=xml:coverage/coverage.xml \
  --cov-report=html:coverage/htmlcov
```
I can get this added to a contribution readme.

**Comment by derjoerg** [comment]
> > This implements the Dimming actuator
> 
> Have you been able to run tests locally? That'll help a lot in triaging the tests.
> 
> Setup a Python virtual environment and install the PyPi package `pip install .` while in the root directory.
> 
> Install Python dev dependencies `pip install src/abbfreeathome/dev-requirements.txt`
> 
> Then run the following to run tests.
> 
> ```
> pytest \                                                                                                                                                                                                                                                
>   --doctest-modules \
>   --junitxml=junit/test-results.xml \
>   --cov=abbfreeathome \
>   --cov-report=xml:coverage/coverage.xml \
>   --cov-report=html:coverage/htmlcov
> ```
> 
> I can get this added to a contribution readme.

Thanks a lot for the explanation. I've set now everything in visual studio code and it seems to run, but ...
My local pytest displays all tests as passed, but here it shows "failed" again.

File: src/abbfreeathome/devices/dimming_actuator.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        """
+
+        value = max(1, value)
+        value = min(value, 100)
+
+        await self._set_brightness_datapoint(str(value))
+        self._brightness = value
+
+        if not self._state:
+            await self.turn_on()
```
[the remark this answers is not in the document]
Comment (derjoerg) [reply src/abbfreeathome/devices/dimming_actuator.py]: Your are right, this works. I just checked which messages came around on the websocket, when I do this through the legacy freeathome-integration.

**Comment by kingsleyadam** [comment]
> Thanks a lot for the explanation. I've set now everything in visual studio code and it seems to run, but ... My local pytest displays all tests as passed, but here it shows "failed" again.
> 
> Additionally, I think set_brightness is not fully tested, but I don't undetstand how to use the "dimming_actuator._api.set_datapoint.assert_called_with()" meaningfull as it returns different stuff as it should 😢

Your latest one is failing because of code coverage. You can view the results under the action summary.

https://github.com/kingsleyadam/local-abbfreeathome/actions/runs/11386836788

Once you run the test it'll publish a code coverage HTML output so you can see what still needs to be covered. It should be at: `./coverage/htmlcov/index.html`

Screenshot 2024-10-17 at 20 12 54

It's just one line where a device update wouldn't affect the state.

**Comment by derjoerg** [comment]
> > Thanks a lot for the explanation. I've set now everything in visual studio code and it seems to run, but ... My local pytest displays all tests as passed, but here it shows "failed" again.
> > Additionally, I think set_brightness is not fully tested, but I don't undetstand how to use the "dimming_actuator._api.set_datapoint.assert_called_with()" meaningfull as it returns different stuff as it should 😢
> 
> Your latest one is failing because of code coverage. You can view the results under the action summary.
> 
> https://github.com/kingsleyadam/local-abbfreeathome/actions/runs/11386836788
> 
> Once you run the test it'll publish a code coverage HTML output so you can see what still needs to be covered. It should be at: `./coverage/htmlcov/index.html`
> 
> Screenshot 2024-10-17 at 20 12 54
> It's just one line where a device update wouldn't affect the state.

Sorry, I'm totally lost here. When I run the pytest-command as stated by you above I get a coverage-output stating that everything is OK.
Also, when I look at your hint I don't understand why this line is a problem. It is exactly the same like e.g. in MovementDetector

**Comment by kingsleyadam** [comment]
> Sorry, I'm totally lost here. When I run the pytest-command as stated by you above I get a coverage-output stating that everything is OK.
> 
> Also, when I look at your hint I don't understand why this line is a problem. It is exactly the same like e.g. in MovementDetector

No worries, I'll take a look tomorrow.

I noticed an issue with how I setup the tests that make it really hard to triage in VS Code. I'll make a change tomorrow which will allow you to see coverage in VS Code and Debug the Test Code including stepping into the module code.

File: tests/test_dimming_actuator.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+def dimming_actuator(mock_api):
+    """Set up the dimming instance for testing the DimmingActuator device."""
+    inputs = {
+        "idp0000": {"pairingID": 1, "value": "0"},
+        "idp0002": {"pairingID": 17, "value": "50"},
+    }
+    outputs = {
+        "odp0000": {"pairingID": 256, "value": "0"},
+        "odp0001": {"pairingID": 272, "value": "50"},
+    }
```
Comment (kingsleyadam) [inline tests/test_dimming_actuator.py:27]: Add an additional output that exists in a dimmer actuator that doesn't affect the state.

```suggestion
       outputs = {
        "odp0000": {"pairingID": 256, "value": "0"},
        "odp0001": {"pairingID": 272, "value": "50"},
        "odp0003": {"pairingID": 257, "value": "0"},
    }
```

File: tests/test_dimming_actuator.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    dimming_actuator._refresh_state_from_output(
+        output={"pairingID": 272, "value": 75},
+    )
+    assert dimming_actuator.brightness == 75
+
+    # Check output that does NOT affect the state.
+    dimming_actuator._refresh_state_from_output(
+        output={"pairingID": 256, "value": "0"},
+    )
+    assert dimming_actuator.state is False
```
Comment (kingsleyadam) [inline tests/test_dimming_actuator.py:116]: Refresh the state from an output that doesn't affect the state.

```suggestion
    # Check output that does NOT affect the state,
    # ensure state, brightness are unchanged.
    dimming_actuator._refresh_state_from_output(
        output={"pairingID": 257, "value": "0"},
    )
    assert dimming_actuator.brightness == 75
    assert dimming_actuator.state is True
```

File: tests/test_dimming_actuator.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
+"""Test class to test the DimmingActuator device."""
+
+from unittest.mock import AsyncMock
+
+import pytest
+
+from abbfreeathome.api import FreeAtHomeApi
+from abbfreeathome.devices.dimming_actuator import DimmingActuator
```
Comment (kingsleyadam) [inline tests/test_dimming_actuator.py:8]: Reference module files directly (vs installed module)

```suggestion
from src.abbfreeathome.api import FreeAtHomeApi
from src.abbfreeathome.devices.dimming_actuator import DimmingActuator
```

**Review by kingsleyadam — CHANGES_REQUESTED** [review]
With these adjustments you should get 100% coverage.

```
Name                                             Stmts   Miss  Cover
--------------------------------------------------------------------
src/__init__.py                                      0      0   100%
src/abbfreeathome/__init__.py                        0      0   100%
src/abbfreeathome/api.py                           150      0   100%
src/abbfreeathome/bin/__init__.py                    0      0   100%
src/abbfreeathome/bin/function.py                  403      0   100%
src/abbfreeathome/bin/interface.py                   8      0   100%
src/abbfreeathome/bin/pairing.py                   323      0   100%
src/abbfreeathome/bin/parameter.py                   0      0   100%
src/abbfreeathome/devices/__init__.py                0      0   100%
src/abbfreeathome/devices/base.py                   70      0   100%
src/abbfreeathome/devices/dimming_actuator.py       47      0   100%
src/abbfreeathome/devices/movement_detector.py      29      0   100%
src/abbfreeathome/devices/switch_actuator.py        32      0   100%
src/abbfreeathome/devices/switch_sensor.py          23      0   100%
src/abbfreeathome/devices/trigger.py                14      0   100%
src/abbfreeathome/exceptions.py                     33      0   100%
src/abbfreeathome/freeathome.py                     65      0   100%
--------------------------------------------------------------------
TOTAL                                             1197      0   100%
```

**Review by kingsleyadam — APPROVED** [review]

**Pull request merged**